[FSDP] Save `_all_handles`; `_all_fsdp_states` to root

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #95507 [FSDP] Expose private prefetch limit setters
* **#95465 [FSDP] Save `_all_handles`; `_all_fsdp_states` to root**
* #95343 [FSDP] Save `_fsdp_states` on root

- The previous PR addressed one tree traversal in `_root_pre_forward()` but not the main one from `_get_fsdp_handles()` that runs for all settings.
- This PR saves `_all_handles` to cache `_get_fsdp_handles()` and `_all_fsdp_states` to cache `_get_fsdp_states()` (renamed from `_fsdp_states` compared to last PR) on the root state.
- This PR introduces a dummy `_RootFSDPState` class that inherits from `_FSDPState` to be used only for type checking since some attributes are only defined for root states.
    - I found this approach to be better than adding `_p_assert(state.root_only_attr is not None, ...)` upon each usage of `root_only_attr`.
    - This hopefully also helps readers to quickly see which attributes are defined only on root states.

Commit:
[FSDP] Save `_all_handles`; `_all_fsdp_states` to root

[ghstack-poisoned]

Commit:
Update on "[FSDP] Save `_all_handles`; `_all_fsdp_states` to root"


- The previous PR addressed one tree traversal in `_root_pre_forward()` but not the main one from `_get_fsdp_handles()` that runs for all settings.
- This PR saves `_all_handles` to cache `_get_fsdp_handles()` and `_all_fsdp_states` to cache `_get_fsdp_states()` (renamed from `_fsdp_states` compared to last PR) on the root state.
- This PR introduces a dummy `_RootFSDPState` class that inherits from `_FSDPState` to be used only for type checking since some attributes are only defined for root states.
    - I found this approach to be better than adding `_p_assert(state.root_only_attr is not None, ...)` upon each usage of `root_only_attr`.
    - This hopefully also helps readers to quickly see which attributes are defined only on root states.

[ghstack-poisoned]

fduwjj reviewed (approved):
LGTM and thanks for the quick change!

fegin commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -57,14 +57,31 @@ def __init__(self) -> None:
+        super().__init__()
+        # Save these static lists to avoid the repeated tree traversals
+        self._all_fsdp_states: List[_FSDPState] = []
+        self._all_handles: List[flat_param_file.FlatParamHandle] = []
```
I feel like it's a little bit convoluted. If `_RootFSDPState` is only for typing purpose and is never instantiated, then the declaration of `_all_fsdp_states` and `_all_handles` are confusing. Basically, this mixes the duck typing with static typing -- we dynamically assign `_all_fsdp_states` to a `_FSDPState` instance and then cast it to `_RootFSDPState` to fulfill the static typing requirement. 

What is the issue of declaring `_all_fsdp_states` and `_all_handles` in `_FSDPState` and leave them to be empty for non-root state?

awgu replied:
Thanks @fegin  for the insight!

What you said makes sense to me. I think leaving `_all_fsdp_states` and `_all_handles` in `_FSDPState` empty for non-root states sounds good to me.

I was originally thinking that in the existing way, we could still get attribute-not-defined errors if we accidentally try to access for non-root states. (For example, this is the existing approach for `_post_backward_callback_queued`, where in @zhaojuanmao 's original implementation, she wanted that to erroring so that we can catch accidental usages.)

awgu replied:
Is it worthwhile to make `_all_fsdp_states` and `_all_handles` into properties that check `_is_root == True` upon access?

awgu replied:
Update: I got rid of `_RootFSDPState`.

Commit:
Update on "[FSDP] Save `_all_handles`; `_all_fsdp_states` to root"


- The previous PR addressed one tree traversal in `_root_pre_forward()` but not the main one from `_get_fsdp_handles()` that runs for all settings.
- This PR saves `_all_handles` to cache `_get_fsdp_handles()` and `_all_fsdp_states` to cache `_get_fsdp_states()` (renamed from `_fsdp_states` compared to last PR) on the root state.
- This PR introduces a dummy `_RootFSDPState` class that inherits from `_FSDPState` to be used only for type checking since some attributes are only defined for root states.
    - I found this approach to be better than adding `_p_assert(state.root_only_attr is not None, ...)` upon each usage of `root_only_attr`.
    - This hopefully also helps readers to quickly see which attributes are defined only on root states.

[ghstack-poisoned]

awgu commented:
@pytorchbot merge
